### PR TITLE
Bar Charts 1. State Vs National park and 2. State Vs all generic parks

### PR DESCRIPTION
**Pull request by rutujapadgilwar**
## Description
This pull request introduces two bar charts: one for comparing the counts of national parks per state and another for comparing the overall park counts per state. The park endpoint was utilized to gather the necessary data for these charts.

## Changes Made
1. I have implemented a single BarChartDataProcessing component that has been created to process the data for the bar chart. This component is reused in both charts to avoid code duplication and promote code reusability.
2. A modal component has been implemented, which is triggered when a user clicks on a column in the bar chart. The modal displays a list of park names for the selected state, along with the corresponding park images. Each park name is a link that redirects the user to the "parkInfoPage".
3. The modal view has been updated to display two parks per row if a state has more than 10 parks, otherwise it shows one park per row
4. The code has been updated to ensure media responsiveness for the bar chart and modal view.

## Screenshots:  
State Vs National Parks
image

State Vs all generic parks
image

Modal View:
park count > 10
image

park count < 10
image

Screen Size < 550px
image

Please review the code changes and provide any further feedback or required modifications.

### CODE REVIEW

File: park-dashboard/src/Components/BarChartStateVsNationalPark/BarChartDataProcessing.jsx [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+  const chartRef = useRef();
+  const handleClick = (event) => {
+    const index = getElementAtEvent(chartRef.current, event)[0]?.index;
+
+    if (index !== undefined) {
+      const stateName = stateNameToAbbreviations[chartData.labels[index]];
+      const parkNames = parkNamesByStates[stateName];
+      setSelectedColumnName(chartData.labels[index]);
+      setselectedStateParkNames(parkNames);
```
Comment (schoure98) [inline park-dashboard/src/Components/BarChartStateVsNationalPark/BarChartDataProcessing.jsx]: Consistent naming might make your code follow the best practices. 
if possible update setselectedStateParkNames to setSelectedStateParkNames.. following the camel case.

File: park-dashboard/src/Components/BarChartStateVsNationalPark/BarChartDataProcessing.jsx [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+      setShowModal(true);
+    } else {
+      console.log("Not clicked on a bar column");
```
Comment (schoure98) [inline park-dashboard/src/Components/BarChartStateVsNationalPark/BarChartDataProcessing.jsx]: Can you remove the console.log?

File: park-dashboard/src/Components/BarChartStateVsNationalPark/BarChartDataProcessing.jsx [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+      <div className="chartDisplay1">
+        <Bar
+          data={chartData}
+          options={options}
+          onClick={handleClick}
+          ref={chartRef}
+        />
+      </div>
+
+      <Modal show={showModal} onHide={handleCloseModal} centered className="PopUpModal">
```
Comment (schoure98) [inline park-dashboard/src/Components/BarChartStateVsNationalPark/BarChartDataProcessing.jsx:188]: Excellent work on the Modal, nice styling.

File: park-dashboard/src/Components/BarChartStateVsNationalPark/BarChartDataProcessing.jsx [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+  const chartData = {
+    labels: barChartData.map(
+      (item) => stateAbbreviation[item.state] || item.state
+    ),
+    datasets: [
+      {
+        data: barChartData.map((item) => item.count),
+        backgroundColor: [
+          "#ff6384",
```
Comment (schoure98) [inline park-dashboard/src/Components/BarChartStateVsNationalPark/BarChartDataProcessing.jsx:72]: Good color selection.

File: park-dashboard/src/Components/BarChartStateVsNationalPark/BarChartDataProcessing.jsx [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+        state = "Other";
+      } else {
+        state = park.state;
+      }
+      counts[state] = (counts[state] || 0) + 1;
+      return counts;
+    }, {});
+
+    const parkCountsArray = Object.entries(parkCounts)
+      .map(([state, count]) => ({
```
Comment (schoure98) [inline park-dashboard/src/Components/BarChartStateVsNationalPark/BarChartDataProcessing.jsx]: stateCounts, parkCounts as variable names might be more easily understandable.

**Review by schoure98 — COMMENTED** [review]
Excellent work on the bar charts and modal component. Loved the styling.

**Comment by juhiaugustine** [comment]
> ## Description
> This pull request introduces two bar charts: one for comparing the counts of national parks per state and another for comparing the overall park counts per state. The park endpoint was utilized to gather the necessary data for these charts.
> 
> ## Changes Made
> 1. I have implemented a single BarChartDataProcessing component that has been created to process the data for the bar chart. This component is reused in both charts to avoid code duplication and promote code reusability.
> 2. A modal component has been implemented, which is triggered when a user clicks on a column in the bar chart. The modal displays a list of park names for the selected state, along with the corresponding park images. Each park name is a link that redirects the user to the "parkInfoPage".
> 3. The modal view has been updated to display two parks per row if a state has more than 10 parks, otherwise it shows one park per row
> 4. The code has been updated to ensure media responsiveness for the bar chart and modal view.
> 
> ## Screenshots:
> State Vs National Parks image
> 
> State Vs all generic parks image
> 
> Modal View: park count > 10 image
> 
> park count < 10 image
> 
> Screen Size < 550px image
> 
> Please review the code changes and provide any further feedback or required modifications.

Wow! Beautiful work on the visualizations and modal component. Excellent color selection. Great work!

File: park-dashboard/src/Components/BarChartStateVsNationalPark/BarChartDataProcessing.jsx (hunk 1 above)
Comment (rutujapadgilwar) [reply park-dashboard/src/Components/BarChartStateVsNationalPark/BarChartDataProcessing.jsx]: Thank you for bringing it to my attention. I have made the necessary update.

File: park-dashboard/src/Components/BarChartStateVsNationalPark/BarChartDataProcessing.jsx (hunk 5 above)
Comment (rutujapadgilwar) [reply park-dashboard/src/Components/BarChartStateVsNationalPark/BarChartDataProcessing.jsx]: Thank you for bringing it to my attention. I have made the necessary update.

File: park-dashboard/src/Components/BarChartStateVsNationalPark/BarChartDataProcessing.jsx (hunk 2 above)
Comment (rutujapadgilwar) [reply park-dashboard/src/Components/BarChartStateVsNationalPark/BarChartDataProcessing.jsx]: console.log() has been removed now.

**Comment by juhiaugustine** [comment]
Everything looks perfect. One small improvisation I could think of is making the Link Style color black. 

Consider using a uniform color scheme throughout the website by changing the link color to black. This will create a more visually consistent and appealing design.

**Comment by rutujapadgilwar** [comment]
> Everything looks perfect. One small improvisation I could think of is making the Link Style color black.
> 
> Consider using a uniform color scheme throughout the website by changing the link color to black. This will create a more visually consistent and appealing design.

Hey Juhi, Thanks for the suggestion. I have added the hover color to the link. The image is attached below:
image

**Review by juhiaugustine — APPROVED** [review]

**Pull request merged**